system/lib/wasmfs/paths.h

```diff
@@ -5,44 +5,71 @@
 
 #pragma once
 
-#include <vector>
+#include <cassert>
+#include <memory>
+#include <string_view>
+#include <variant>
 
 #include "file.h"
 
-namespace wasmfs {
+namespace wasmfs::path {
 
-struct ParsedPath {
-  std::optional<Directory::Handle> parent;
-  std::shared_ptr<File> child;
+// Typically -ENOTDIR or -ENOENT.
+using Error = long;
+
+// The parent directory and the name of an entry within it.
+using ParentChild = std::pair<std::shared_ptr<Directory>, std::string_view>;
+
+struct ParsedParent {
+private:
+  std::variant<Error, ParentChild> val;
+
+public:
+  ParsedParent(Error err) : val(err) {}
+  ParsedParent(ParentChild pair) : val(pair) {}
+  // Always ok to call, returns 0 if there is no error.
+  long getError() {
+    if (auto* err = std::get_if<Error>(&val)) {
+      assert(*err != 0 && "Unexpected zero error value");
```

Comment:
if we use `__wasi_errno_t` then this could replace 0 with `__WASI_ERRNO_SUCCESS`

Reply:
Maybe we should standardize our use of __wasi_errno_t and __WASI_ERRNO_SUCCESS in a follow-up PR. We use `long` and 0 at least as much.